### PR TITLE
feat: add ResponseFormat support to call_model_input_filter (#3563)

### LINKED ISSUE #3563: Add custom `ResponseFormat` support in CallModelInputFilter callback
`call_model_input_filter` of RunConfig  is invoked immediately before calling the model which allows you to edit the input sent to the model, Is there any change to add `response_format: ResponseFormat | None` field to support modify response format before calling the model.

### PR DESCRIPTION
**Pull request by ylzn567**
Closes #3563

### Description
This PR adds `response_format` support to the `call_model_input_filter` callback in `RunConfig`. This allows developers to dynamically modify the expected response format (e.g., enforcing a strict JSON schema) immediately before the model is called.

### Changes Made
* Added `response_format: ResponseFormat | None` to the `ModelInputData` dataclass in `src/agents/run_config.py`.
* Updated the `maybe_filter_model_input` function in `src/agents/run_internal/turn_preparation.py` to accept and pass the `response_format`.
* Updated the run loop orchestration in `src/agents/run_internal/run_loop.py` to correctly inject `model_settings.response_format` into the filter during both streamed and non-streamed runs.

### Next Steps
* I'm opening this PR to get early feedback on the core implementation.
* **WIP:** I am currently working on adding the contract tests requested in the issue to verify the durability around this runtime boundary. I will push those commits shortly when I have more time to dedicate to them. 

*(If you have a specific existing test fixture or file you recommend I use as a reference for testing this serialization/resume behavior, I'd appreciate the pointer!)*

### CODE REVIEW

File: src/agents/run_internal/run_loop.py [hunk 1]
```diff
@@ -1815,19 +1820,22 @@ async def get_new_response(
...
+    model_settings = get_model_settings(execution_agent, run_config)
+    model_settings = maybe_reset_tool_choice(public_agent, tool_use_tracker, model_settings)
+
     filtered = await maybe_filter_model_input(
         agent=public_agent,
         run_config=run_config,
         context_wrapper=context_wrapper,
         input_items=input,
         system_instructions=system_prompt,
+        response_format=None,
```
[the remark this answers is not in the document]
Comment (ylzn567) [reply src/agents/run_internal/run_loop.py]: @codex address that feedback

File: src/agents/run_internal/run_loop.py [hunk 2]
```diff
@@ -1373,6 +1374,7 @@ def _tool_search_fingerprint(raw_item: Any) -> str:
         context_wrapper=context_wrapper,
         input_items=input,
         system_instructions=system_prompt,
+        response_format=model_settings.response_format,
```
[the remark this answers is not in the document]
Comment (ylzn567) [reply src/agents/run_internal/run_loop.py]: @codex address that feedback

**Comment by ylzn567** [comment]
Hi team! 👋 
I've implemented the logic to pass and handle `response_format` via the `call_model_input_filter`. However, the Codex bot and I have hit a bit of a Catch-22:

1. If I don't pass `filtered.response_format` to `model.get_response(...)`, Codex flags it as an unused variable (since it drops before the model request).
2. If I *do* pass it, Codex rightly points out that the base `Model` interface (`src/agents/models/interface.py`) and its implementations don't currently accept `response_format` as a kwarg, resulting in a `TypeError`.

To cleanly resolve this, the `Model` interface and all its concrete implementations (ChatCompletion, LiteLLM, Mocks, etc.) would need to be updated to accept this new parameter. 

Since updating the entire model interface layer expands the scope of this PR significantly, I've temporarily reverted the kwarg injection. I would love your architectural guidance:
Would you like me to expand this PR to update `interface.py` and all underlying model wrappers? Or would you prefer to handle the interface expansion separately?

File: src/agents/run_internal/run_loop.py [hunk 3]
```diff
@@ -1893,6 +1923,7 @@ async def rewind_model_request() -> None:
             previous_response_id=previous_response_id,
             conversation_id=conversation_id,
             prompt=prompt_config,
+            response_format=filtered.response_format,
```
[the remark this answers is not in the document]
Comment (ylzn567) [reply src/agents/run_internal/run_loop.py]: @codex address that feedback

**Comment by ylzn567** [comment]
I see @jordanchendev has already provided a complete solution along with the requested contract tests over in #3571! 

Closing this PR to keep the tracker tidy. It was a great learning experience diving into the run loop and input filter architecture. Keep up the great work!
